comm/src/main/scala/coop/rchain/comm/transport/GrpcTransportClient.scala

```diff
@@ -0,0 +1,231 @@
+              GrpcTransport.stream(peer, Blob(sender, packet), messageSize)
+            ).flatMap {
+              case Left(error) =>
+                log.error(s"Error while streaming packet, error: $error") *> delay(
+                  handle(retryCount - 1)
+                )
+              case Right(_) =>
+                log.info(s"Streamed packet $path to $peer")
+            }
+          case Left(error) => log.error(s"Error while streaming packet, error: $error")
+        } else log.debug(s"Giving up on streaming packet $path to $peer")
+
+    handle(retries)
+  }
+
+  private def innerBroadcast(peers: Seq[PeerNode], msg: Protocol): Task[Seq[CommErr[Unit]]] =
+    Task.gatherUnordered(peers.map(send(_, msg, 500.milliseconds, enforce = true)))
+
+  def start(): Task[Unit] = {
+
+    def initQueue(maybeQueue: Option[Cancelable])(create: Task[Cancelable]): Task[Cancelable] =
+      maybeQueue.fold(create) {
+        kp(
+          Task.raiseError[Cancelable](
+            new RuntimeException("TransportLayer server is already started")
+          )
+        )
+      }
+
+    cache.modify { s =>
+      val parallelism = Math.max(Runtime.getRuntime.availableProcessors(), 2)
+      val queueScheduler =
+        Scheduler.fixedPool("tl-dispatcher-client", parallelism, reporter = UncaughtExceptionLogger)
+      for {
+        clientQueue <- initQueue(s.clientQueue) {
+                        import coop.rchain.shared.PathOps.PathDelete
+                        Task.delay {
+                          streamObservable
+                            .flatMap { s =>
+                              Observable
+                                .fromIterable(s.peers)
+                                .mapParallelUnordered(parallelism)(
+                                  streamBlobFile(s.path, _, s.sender, maxMessageSize)
+                                )
+                                .guarantee(s.path.deleteSingleFile[Task]())
+                            }
+                            .subscribe()(queueScheduler)
+                        }
+                      }
+      } yield s.copy(server = None, clientQueue = Some(clientQueue))
+    }
+  }
+
+  def shutdown(msg: Protocol): Task[Unit] = {
+    def shutdownQueue: Task[Unit] = cache.modify { s =>
+      for {
+        _ <- log.info("Shutting down transport layer")
+        _ <- s.clientQueue.fold(Task.unit)(q => Task.delay(q.cancel()))
+      } yield s.copy(server = None, shutdown = true)
+    }
+
+    def sendShutdownMessages: Task[Unit] =
+      for {
+        peers <- cache.read.map(_.connections.keys.toSeq)
+        _     <- log.info("Sending shutdown message to all peers")
+        _     <- innerBroadcast(peers, msg)
+        _     <- log.info("Disconnecting from all peers")
+        _     <- Task.gatherUnordered(peers.map(disconnect))
+      } yield ()
+
+    cache.read.flatMap { s =>
+      if (s.shutdown) Task.unit
+      else shutdownQueue *> sendShutdownMessages
```

Comment:
use `>>` instead 
1. stack saftey
2. order of execution!

Reply:
You made my day! :)